Path utils refactor

RussKie opened the pull request:
As discussed in https://github.com/gitextensions/gitextensions/pull/4032#issuecomment-333517989

* Deprecate PathUtil.DirectoryExists
* Deprecate PathUtil.PathExists
* Move PathEqualityComparer to outer scope

jbialobr commented on GitCommands/PathUtil.cs:
```diff
@@ -188,36 +156,9 @@ public static bool IsValidPath(string aPath)
             return fi != null;
         }
                 
-        public static bool PathExists(string aPath)
```
These methods was created to return false for cases when the given path is of incorrect form (causing exceptions being caught - lines 198-200). The uses case is for expanding to a full path based on environment PATH variable. We are seeking only for valid paths and don't want to propagate exceptions to UI.

RussKie commented:
Thanks, I'll make sure exceptions here are contained

On 09/10/2017 12:06 AM, "Janusz Białobrzewski" <notifications@github.com>
wrote:

> *@jbialobr* commented on this pull request.
> ------------------------------
>
> In GitCommands/PathUtil.cs
> <https://github.com/gitextensions/gitextensions/pull/4046#discussion_r143354144>
> :
>
> > @@ -188,36 +156,9 @@ public static bool IsValidPath(string aPath)
>              return fi != null;
>          }
>
> -        public static bool PathExists(string aPath)
>
> These methods was created to return false for cases when the given path is
> of incorrect form (causing exceptions being caught - lines 198-200). The
> uses case is for expanding to a full path based on environment PATH
> variable. We are seeking only for valid paths and don't want to propagate
> exceptions to UI.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4046#pullrequestreview-67856833>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXsRcuYfwseYRlISD5jpS2R9UDvVUks5sqMjvgaJpZM4PxsYj>
> .
>

RussKie commented:
`File.Exists` doesn't throw, but I added few unit tests just in case

jbialobr commented:
The changes are not semantically equal. The current implementation stops the seeking process after the first exception being thrown. The previous implementation kept seeking after an exception being thrown.

RussKie commented:
That is actually not true - they are equal.
I've added unit tests before I've wrapped the implementations in try/catch. 

There are two reasons why I've added try/catch:
1. passing `null` causes `File.Exists` to throw, and it would throw in the original implementation as well
2. `TryXxx` methods should not throw in any case

`Environment` needs to be abstracted so `Environment.Get/SetEnvironmentVariable` can be tested, then more tests can be written to test this functionality

jbialobr commented:
```csharp
                foreach (var path in GetEnvironmentValidPaths())
                {
                    fullPath = Path.Combine(path, fileName);
the created fullPath can be too long giving PathTooLongException
                    if (File.Exists(fullPath))
                    {
                        return true;
                    }
                }
```

RussKie commented:
![image](https://user-images.githubusercontent.com/4403806/31382981-257cfe04-ae05-11e7-803f-b5b83610454b.png)

RussKie commented:
That works on both W8 and W10

jbialobr commented:
Yes, you are right. Now I see that the root cause for #2924 was `Path.Combine`, not `Path.Exists` :)
